File: sliding_window/#3_longest_substring_without_repeating_characters/lengthOfLongestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  /**
   * 遍歷s, 並且用map紀錄char [char, count]
   * 當出現重複char時, 要把start移動到該重複char右邊的char, 同時要把start走過的char的count減掉
   */
  let start = 0,
    end = 0,
    maxLen = 0;
  const map = new Map();

  while (end < s.length) {
    if (map.get(s[end]) > 0) {
      map.set(s[end], map.get(s[end]) + 1);
    } else {
      map.set(s[end], 1);
    }

    let count = map.get(s[end]);

    while (count > 1) {
      startChar = s[start];
      charCount = map.get(startChar);
      map.set(startChar, charCount - 1);
      count = map.get(s[end]);
      start++;
    }

    maxLen = Math.max(maxLen, end - start + 1);
    end++;
  }

  console.log(maxLen);
  return maxLen;
};

lengthOfLongestSubstring("abcabcbb"); // 3
lengthOfLongestSubstring("bbbbb"); // 1
lengthOfLongestSubstring("pwwkew"); // 3
lengthOfLongestSubstring(""); // 0
lengthOfLongestSubstring(" "); // 1

File: sliding_window/#3_longest_substring_without_repeating_characters/lengthOfLongestSubstring.go
package main

import "fmt"

func main() {
	lengthOfLongestSubstring("abcabcbb")
	lengthOfLongestSubstring("bbbbb")
	lengthOfLongestSubstring("pwwkew")
	lengthOfLongestSubstring("abba")
	lengthOfLongestSubstring("")  // 0
	lengthOfLongestSubstring(" ") // 1
}

func lengthOfLongestSubstring(s string) int {
	start, maxLen := -1, 0
	hash := make(map[rune]int)

	for i, ch := range s {
		if _, ok := hash[ch]; ok && hash[ch] > start {
			/**
			* hash[ch] > start 用於判斷目前 sliding window的start位置是否超過hash[ch]的index
			* 以 "abba" 當遇到重複元素b時, start將會移動到 index 0, 遇到重複a時需檢查hash['a']的index有沒有大於start, 否則start會往回走
			 */
			start = hash[ch]
			fmt.Println("ok", start)
		}
		hash[ch] = i

		maxLen = max(maxLen, i-start)
	}

	fmt.Println(maxLen)
	return maxLen
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// https://leetcode.com/problems/longest-substring-without-repeating-characters/solutions/3189769/go-simple-o-n-solution-with-comments/
// func lengthOfLongestSubstring(s string) (max int) {
// 	symbols := map[byte]int{} // value is latest seen place of symbol
// 	length := 0               // length of the current substring
// 	for i := 0; i != len(s); i++ {
// 		v, ok := symbols[s[i]]
// 		symbols[s[i]] = i
// 		if ok && v >= i-length { //s[i] in current substring
// 			length = i - v // change substring to s[v+1]...s[i] ("abca"->"bca")
// 		} else {
// 			length += 1
// 		}
// 		if length > max {
// 			max = length
// 		}
// 	}
// 	fmt.Println(max)
// 	return
// }
